packages/salesforcedx-apex-debugger/src/adapter/apexDebug.ts

```diff
@@ -304,38 +324,55 @@ export class ObjectReferenceContainer implements VariableContainer {
         return new ApexVariable(
           value,
           ApexVariableKind.Field,
-          variableReference
+          variableReference,
+          session.getNumOfChildren(variableReference)
         );
       })
     );
   }
+
+  public getNumberOfChildren(): number | undefined {
+    return this.size;
+  }
 }
 
 export class CollectionReferenceContainer extends ObjectReferenceContainer {
   public async expand(
     session: ApexDebug,
     filter: FilterType,
-    start: number,
-    count: number
+    start?: number,
+    count?: number
   ): Promise<ApexVariable[]> {
     if (!this.reference.value) {
       // this object is empty
       return [];
     }
-
-    return Promise.all(
-      this.reference.value.map(async value => {
-        const variableReference = await session.resolveApexIdToVariableReference(
-          this.requestId,
-          value.ref
-        );
-        return new ApexVariable(
-          value,
+    if (start === undefined) {
+      start = 0;
+    }
+    if (count === undefined) {
+      count = this.reference.value.length;
+    }
+    const apexVariables: ApexVariable[] = [];
+    for (
+      let i = start;
+      i < start + count && i < this.reference.value.length;
```

Comment:
Previously, it would fetched every variable in the collection. Now, we fetch from the provided `start` until `start + count`. Those values are provided by variablesRequest() which is a request for a variable from VS Code to the adapter. VS Code knows whether or not a variable has paginated children, so it will automatically provide those start and count values depending on which page the user expands.